crates/edr_napi/src/solidity_tests/config.rs

```diff
@@ -1,21 +1,697 @@
-use std::{collections::HashMap, path::PathBuf};
+use std::{collections::HashMap, fmt::Debug, path::PathBuf};
 
-use alloy_primitives::{address, Address};
+use alloy_primitives::{address, hex, Address, B256};
 use edr_eth::U256;
 use forge::{
+    fork::CreateFork,
     inspectors::cheatcodes::CheatsConfigOptions,
     opts::{Env as EvmEnv, EvmOpts},
 };
-use foundry_config::{
-    cache::StorageCachingConfig, FsPermissions, FuzzConfig, GasLimit, InvariantConfig, RpcEndpoint,
-    RpcEndpoints,
+use foundry_config::{FsPermissions, FuzzConfig, InvariantConfig, RpcEndpoint, RpcEndpoints};
+use napi::{
+    bindgen_prelude::{BigInt, Buffer},
+    Either, Status,
 };
+use napi_derive::napi;
+
+use crate::cast::TryCast;
+
+/// Default address of `tx.origin` in Foundry
+///
+/// `0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38`
+const DEFAULT_SENDER: Address = address!("1804c8AB1F12E6bbf3894d4083f33e07309d1f38");
+
+/// Solidity test runner configuration arguments exposed through the ffi.
+/// Docs based on https://book.getfoundry.sh/reference/config/testing
+#[napi(object)]
+#[derive(Clone)]
+pub struct SolidityTestRunnerConfigArgs {
+    /// The absolute path to the project root directory.
+    /// Relative paths in cheat codes are resolved against this path.
+    pub project_root: String,
+    /// Configures the permissions of cheat codes that access the file system.
+    pub fs_permissions: Option<Vec<PathPermission>>,
+    /// Whether to collect traces. Defaults to false.
+    pub trace: Option<bool>,
+    /// Whether to collect debug info. Defaults to false.
+    pub debug: Option<bool>,
+    /// Address labels for traces. Defaults to none.
+    pub labels: Option<Vec<AddressLabel>>,
+    /// Whether to enable isolation of calls. In isolation mode all top-level
+    /// calls are executed as a separate transaction in a separate EVM
+    /// context, enabling more precise gas accounting and transaction state
+    /// changes.
+    /// Defaults to false.
+    pub isolate: Option<bool>,
+    /// Whether or not to enable the ffi cheatcode.
+    /// Warning: Enabling this cheatcode has security implications, as it allows
+    /// tests to execute arbitrary programs on your computer.
+    /// Defaults to false.
+    pub ffi: Option<bool>,
+    /// The value of `msg.sender` in tests as hex string.
+    /// Defaults to `0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38`.
+    pub sender: Option<Buffer>,
+    /// The value of `tx.origin` in tests as hex string.
+    /// Defaults to `0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38`.
+    pub tx_origin: Option<Buffer>,
```

Comment:
What does origin mean in this context?

I see that there is a separate `sender` config option, so it must be something distinct 🤔

Reply:
`origin` and `sender` are different things in the EVM.

Let's say that you use account `A` to call contract `Foo`, which in turn calls contract `Bar`. In the context of `Bar`'s execution, the sender is `Foo`, but the origin of the transaction is `A`.

In solidity, these values can be obtained with `tx.origin` and `msg.sender`. In the EVM, the opcodes are `0x32 ORIGIN` and `0x33 CALLER`.

Reply:
Thanks!